ACPs/118-warp-signature-request/README.md

```diff
@@ -0,0 +1,88 @@
+
+## Motivation
+
+Warp message signatures consist of an aggregate BLS signature composed of the individual signatures of a subnet's validators. Individual signatures need to be retreivable by the party that wishes to construct an aggregate signature. At present, this is left to VMs to implement, as is the case with [Subnet EVM](https://github.com/ava-labs/subnet-evm/blob/v0.6.7/plugin/evm/message/signature_request.go#20) and [Coreth](https://github.com/ava-labs/coreth/blob/v0.13.6-rc.0/plugin/evm/message/signature_request.go#L20)
+
+This creates friction in applications that are intended to operate across many VMs (or distinct implementations of the same VM). As an example, the reference Warp message relayer implementation, [awm-relayer](https://github.com/ava-labs/awm-relayer), fetches individual signatures from validators and aggregates them before sending the Warp message to its destination chain for verification. However, Subnet EVM and Coreth have distinct codecs, requiring the relayer to [switch](https://github.com/ava-labs/awm-relayer/blob/v1.4.0-rc.0/relayer/application_relayer.go#L372) according to the target codebase.
+
+Another example is ACP-75, which aims to implement acceptance proofs using Warp. The signature aggregation mechanism is not [specified](https://github.com/avalanche-foundation/ACPs/blob/main/ACPs/75-acceptance-proofs/README.md#signature-aggregation), which is a blocker for that ACP to be marked implementable.
+
+Standardizing the Warp Signature Request interface by defining it as a format for `AppRequest` message payloads in AvalancheGo would simplify the implementation of ACP-75, and streamline signature aggregation for out-of-protocol services such as Warp message relayers.
+
+## Specification
+
+We propose the following types, implemented as Protobuf types that may be decoded from the `AppRequest`/`AppResponse` `app_bytes` field. By way of example, this approach is currently used to [implement](https://github.com/ava-labs/avalanchego/blob/v1.11.10-status-removal/proto/sdk/sdk.proto#7) and [parse](https://github.com/ava-labs/avalanchego/blob/v1.11.10-status-removal/network/p2p/gossip/message.go#22) gossip `AppRequest` types.
+
+- `SignatureRequest` includes two fields. `data` specifies the payload that the returned signature should correspond to, namely a serialized unsigned Warp message. `justification` is an optional field that specifies arbitrary data that the requested node may use to decide whether or not it is willing to sign `data`.
+
+    ```protobuf
+    message SignatureRequest struct {
+        bytes data = 1;
+        bytes justification = 2;
```

Comment:
How will the `justification` field be used?

If the justification is going to provide a hint like a blockID/messageID, should this data be typed and standardized as well?

Reply:
Could we also be more specific about what it means for this to be an optional field? I think it's better to be more prescriptive for how VMs should use this.

Ie. if the justification includes a blockID, does the VM only serve the request if the justification is provided or is it responsible for parsing the message and figuring out what should have been there if it's not provided?

If we're very explicit upfront, then I think this will make it simpler for VM implementers and integrators that can use a uniform standard rather than needing to adjust to any ambiguity that is left up to the VM to decide.

Reply:
> should this data be typed and standardized as well?

It's completely up to the VM to interpret the `justification` in the context of the Warp payload format contained in `data`. Both `data` and `justification` could be parsed into concrete types by the VM, but I don't think we can specify a type for `justification` in `SignatureRequest` without losing generality.

> Could we also be more specific about what it means for this to be an optional field?

Absolutely. I think an illustrative example showing how `data` and `justification` should resolve to coupled types on the VM side could be useful. I'll put that together.

Reply:
I expanded the description of the fields in the Specification to make it clear that it is the VM's responsibility to define the requirements for `data` and `justification`. I also clarified the `Confirm that an event did not occur` use case to make the example more concrete.